compile error for VS14(Visual Studio 2015)

```bash
R:\pyconcrete>E:\MicrosoftVisualStudio14.0\VC\bin\vcvars32.bat
R:\pyconcrete>SET VS100COMNTOOLS=%VS140COMNTOOLS%

R:\pyconcrete>set DISTUTILS_USE_SDK=1

R:\pyconcrete>set MSSdk=1

R:\pyconcrete>py3 setup.py build --compiler=msvc

R:\pyconcrete>e:\prg\py\Anaconda3_64\python.exe setup.py build --compiler=msvc
running build
running build_py
running build_ext
building 'pyconcrete._pyconcrete' extension
E:\MicrosoftVisualStudio14.0\VC\BIN\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG
-Isrc\pyconcrete_ext\openaes\inc -Isrc\pyconcrete_ext\include_win -Ie:\prg\py\An
aconda3_64\include -Ie:\prg\py\Anaconda3_64\include /Tcsrc\pyconcrete_ext\pyconc
rete.c /Fobuild\temp.win-amd64-3.4\Release\src\pyconcrete_ext\pyconcrete.obj
pyconcrete.c
src\pyconcrete_ext\openaes\inc\oaes_lib.h(84): error C2061: 语法错误: 标识符“OA
ES_OPTION”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(84): error C2059: 语法错误:“;”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(126): error C2146: 语法错误: 缺少“)”
(在标识符“option”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(126): error C2081: “OAES_OPTION”: 形
参表中的名称非法
src\pyconcrete_ext\openaes\inc\oaes_lib.h(126): error C2061: 语法错误: 标识符“o
ption”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(126): error C2059: 语法错误:“;”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(126): error C2059: 语法错误:“,”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(126): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): error C2081: “uint8_t”: 形参表
中的名称非法
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): error C2143: 语法错误: 缺少“;”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(137): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): error C2081: “uint8_t”: 形参表
中的名称非法
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): error C2143: 语法错误: 缺少“;”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(142): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): error C2146: 语法错误: 缺少“;”
(在标识符“data_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(146): error C2040: “data_len”:“int
”与“int *”的间接寻址级别不同
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): error C2146: 语法错误: 缺少“;”
(在标识符“data_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(150): error C2040: “data_len”:“int
”与“int *”的间接寻址级别不同
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): error C2146: 语法错误: 缺少“;”
(在标识符“m_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): error C2143: 语法错误: 缺少“;”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(154): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2146: 语法错误: 缺少“;”
(在标识符“c_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2040: “c_len”:“int”与
“int *”的间接寻址级别不同
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2143: 语法错误: 缺少“;”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(158): error C2040: “m_len”:“int *”
与“int”的间接寻址级别不同
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2081: “uint8_t”: 形参表
中的名称非法
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2146: 语法错误: 缺少“;”
(在标识符“c_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(161): error C2040: “c_len”:“int”与
“int *”的间接寻址级别不同
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2081: “uint8_t”: 形参表
中的名称非法
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2146: 语法错误: 缺少“;”
(在标识符“c_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(163): error C2040: “c_len”:“int”与
“int *”的间接寻址级别不同
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): error C2143: 语法错误: 缺少“)”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): error C2143: 语法错误: 缺少“{”
(在“*”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): warning C4142: "size_t": 类型的
良性重定义
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): error C2370: “size_t”: 重定义
；不同的存储类
E:\MicrosoftVisualStudio14.0\VC\INCLUDE\vcruntime.h(198): note: 参见“size_t”的
声明
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): error C2146: 语法错误: 缺少“;”
(在标识符“data_len”的前面)
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): error C2059: 语法错误:“)”
src\pyconcrete_ext\openaes\inc\oaes_lib.h(167): error C2040: “data_len”:“int
”与“int *”的间接寻址级别不同
src\pyconcrete_ext\pyconcrete.c(38): warning C4476: "printf": 格式说明符中的类型
字段字符“P”未知
src\pyconcrete_ext\pyconcrete.c(38): warning C4477: “printf”: 格式字符串“%d”
需要类型“int”的参数，但可变参数 1 拥有了类型“unsigned char *”
src\pyconcrete_ext\pyconcrete.c(38): warning C4474: printf: 格式字符串中传递的参
数太多
src\pyconcrete_ext\pyconcrete.c(38): note: 占位符和其参数预计 1 可变参数，但提供
的却是 2 参数
src\pyconcrete_ext\pyconcrete.c(48): warning C4013: “oaes_key_import_data”未定
义；假设外部返回 int
src\pyconcrete_ext\pyconcrete.c(99): warning C4013: “oaes_encrypt_block”未定义
；假设外部返回 int
src\pyconcrete_ext\pyconcrete.c(79): warning C4996: 'fopen': This function or va
riable may be unsafe. Consider using fopen_s instead. To disable deprecation, us
e _CRT_SECURE_NO_WARNINGS. See online help for details.
C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt\stdio.h(205): n
ote: 参见“fopen”的声明
src\pyconcrete_ext\pyconcrete.c(86): warning C4996: 'fopen': This function or va
riable may be unsafe. Consider using fopen_s instead. To disable deprecation, us
e _CRT_SECURE_NO_WARNINGS. See online help for details.
C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt\stdio.h(205): n
ote: 参见“fopen”的声明
src\pyconcrete_ext\pyconcrete.c(205): warning C4013: “oaes_decrypt_block”未定
义；假设外部返回 int
error: command 'E:\\MicrosoftVisualStudio14.0\\VC\\BIN\\cl.exe' failed with exit
 status 2
```

--- a/README.md
+++ b/README.md
@@ -111,6 +111,35 @@ $ TEST_PYE_PERFORMANCE_COUNT=1 ./pyconcrete-admin.py test
 
 Building on Windows
 --------------
+#### Python 2.7 - Visual Studio 2008
+https://www.microsoft.com/en-us/download/details.aspx?id=44266
+
+* Open VS2008 Command Prompt
+* `set DISTUTILS_USE_SDK=1`
+* `set SET MSSdk=1`
+* create `distutils.cfg` and put inside
+    ```text
+    [build]
+    compiler=msvc
+    ```
+
+#### Python 3.5, 3.6 - Visual Studio 2015
+
+[MSVC 2015 Build Tools](http://landinghub.visualstudio.com/visual-cpp-build-tools)
+
+[Document](https://matthew-brett.github.io/pydagogue/python_msvc.html#python-3-5-3-6)
+
+* make sure setuptools >= 24.0
+    ```sh
+    python -c 'import setuptools; print(setuptools.__version__)'
+    ```
+
+* Open VS2015 Build Tools Command Prompt
+* `set DISTUTILS_USE_SDK=1 `
+* `setenv /x64 /release`  or `setenv /x86 /release`
+
+
+#### Reference
 https://matthew-brett.github.io/pydagogue/python_msvc.html
 https://github.com/cython/cython/wiki/CythonExtensionsOnWindows
 

--- a/setup.py
+++ b/setup.py
@@ -59,6 +59,10 @@ def is_mingw():
     return False
 
 
+def is_msvc():
+    return sys.platform == 'win32' and not is_mingw()
+
+
 def hash_key(key):
     if PY2:
         factor = sum([ord(s) for s in key])
@@ -256,9 +260,14 @@ class InstallEx(CmdBase, install):
         print('creating %s' % filename)
 
     def install_exe(self):
-        # install `pyconcrete` to /usr/local/bin
-        self.copy_file(os.path.join(self.build_scripts, 'pyconcrete'),
-                       os.path.join(self.install_scripts, 'pyconcrete'))
+        if sys.platform == 'win32':
+            # install `pyconcrete.exe` to %PYTHON%/Scripts
+            exe_name = 'pyconcrete.exe'
+        else:
+            # install `pyconcrete` to /usr/local/bin
+            exe_name = 'pyconcrete'
+        self.copy_file(os.path.join(self.build_scripts, exe_name),
+                       os.path.join(self.install_scripts, exe_name))
 
 # ================================================= test command ================================================= #
 
@@ -281,28 +290,47 @@ class TestEx(Command):
 
 # ================================================= extension ================================================= #
 
-openaes_include_dirs = [
-    join(EXT_SRC_DIR),
-    join(EXT_SRC_DIR, 'openaes', 'inc'),
-]
-if sys.platform == 'win32' and not is_mingw():
-    openaes_include_dirs.append(join(EXT_SRC_DIR, 'include_win'))
+def get_include_dirs():
+    openaes_include_dirs = [
+        join(EXT_SRC_DIR),
+        join(EXT_SRC_DIR, 'openaes', 'inc'),
+    ]
+    if is_msvc() and PY2:  # Only Python 2.7 & 3.2 Need VisualStudio 2008 (without stdint.h)
+        openaes_include_dirs.append(join(EXT_SRC_DIR, 'include_win'))
+    return openaes_include_dirs
 
+
+def get_libraries(include_python_lib=False):
+    libraries = []
+    if sys.version_info[0] == 3 and sys.version_info[1] >= 5 and is_msvc():
+        # https://stackoverflow.com/questions/32418766/c-unresolved-external-symbol-sprintf-and-sscanf-in-visual-studio-2015
+        libraries = ['legacy_stdio_definitions']
+
+    if is_msvc():
+        link_py = 'python{0}{1}'.format(*sys.version_info[0:2])
+    else:
+        if PY2:
+            link_py_fmt = 'python{version}'
+        else:
+            link_py_fmt = 'python{version}m'
+        link_py = link_py_fmt.format(version=sysconfig.get_python_version())
+
+    if include_python_lib:
+        libraries.append(link_py)
+
+    return libraries
+
+
+include_dirs = get_include_dirs()
 openaes_sources = [
     join(EXT_SRC_DIR, 'openaes', 'src', 'oaes_base64.c'),
     join(EXT_SRC_DIR, 'openaes', 'src', 'oaes_lib.c'),
 ]
 
-if PY2:
-    link_py = 'python{version}'
-else:
-    link_py = 'python{version}m'
-link_py = link_py.format(version=sysconfig.get_python_version())
-
-
 ext_module = Extension(
     'pyconcrete._pyconcrete',
-    include_dirs=openaes_include_dirs,
+    include_dirs=include_dirs,
+    libraries=get_libraries(include_python_lib=False),
     sources=[
         join(EXT_SRC_DIR, 'pyconcrete.c'),
         join(EXT_SRC_DIR, 'pyconcrete_module.c'),
@@ -311,8 +339,8 @@ ext_module = Extension(
 
 exe_module = Extension(
     'pyconcrete',
-    include_dirs=openaes_include_dirs,
-    extra_link_args=['-l'+link_py],
+    include_dirs=include_dirs,
+    libraries=get_libraries(include_python_lib=True),
     sources=[
         join(EXE_SRC_DIR, 'pyconcrete_exe.c'),
         join(EXT_SRC_DIR, 'pyconcrete.c'),

--- a/test/base.py
+++ b/test/base.py
@@ -87,7 +87,8 @@ def build_tmp_pyconcrete(passphrase):
 
         copy_pyconcrete_ext(tmp_dir)
 
-        tmp_pyconcrete_exe = join(tmp_dir, 'scripts', 'pyconcrete')
+        exe_name = 'pyconcrete.exe' if sys.platform == 'win32' else 'pyconcrete'
+        tmp_pyconcrete_exe = join(tmp_dir, 'scripts', exe_name)
         tmp_pyconcrete_dir = tmp_dir
         print('build tmp pyconcrete at "%s"' % tmp_dir)
 

--- a/test/test_exe.py
+++ b/test/test_exe.py
@@ -45,6 +45,7 @@ class TestPyconcreteExe(base.TestPyConcreteBase):
 
         output = subprocess.check_output([self._pyconcrete_exe, pye])
         output = output.decode('utf8')
+        output = output.replace('\r\n', '\n')
         paths = output.split('\n')
 
         self.assertTrue(pye_dir in paths, "pye dir(%s) not in output paths %s" % (pye_dir, paths))

